Firestore/core/src/firebase/firestore/remote/remote_objc_bridge.mm

```diff
@@ -90,33 +84,26 @@
   }
 }
 
-grpc::ByteBuffer ConvertToByteBuffer(NSData* data) {
-  grpc::Slice slice{[data bytes], [data length]};
+template <typename T>
+grpc::ByteBuffer ConvertToByteBuffer(const pb_field_t* fields,
+                                     const T& request) {
+  ByteStringWriter writer;
+  writer.WriteNanopbMessage(fields, &request);
+  ByteString bytes = writer.Release();
+
+  grpc::Slice slice{bytes.data(), bytes.size()};
   return grpc::ByteBuffer{&slice, 1};
 }
 
-// Note: `StatusOr` cannot be used with ARC-managed objects.
-template <typename Proto>
-Proto* ToProto(const grpc::ByteBuffer& message, Status* out_status) {
-  NSError* error = nil;
-  NSData* data = ConvertToNsData(message, &error);
-  if (!error) {
-    Proto* proto = [Proto parseFromData:data error:&error];
-    if (!error) {
-      *out_status = Status::OK();
-      return proto;
-    }
-  }
-
-  std::string error_description =
-      StringFormat("Unable to parse response from the server.\n"
-                   "Underlying error: %s\n"
-                   "Expected class: %s\n"
-                   "Received value: %s\n",
-                   error, [Proto class], ToHexString(message));
-
-  *out_status = {Error::Internal, error_description};
-  return nil;
+template <typename T, typename U>
+std::string DescribeRequest(const pb_field_t* fields, const U& request) {
```

Comment:
Could this take `const Message<U> &`? In addition to making it possible to reuse `CreateByteBuffer` it would make it possible to get the fields out of the message too.